{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "initial_id",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-15T14:26:37.473823Z",
     "start_time": "2025-11-15T14:26:36.957896Z"
    },
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Project root: /home/nivd1/TAUSO\n"
     ]
    }
   ],
   "source": [
    "# conda activate aso_design\n",
    "import sys\n",
    "from pathlib import Path\n",
    "\n",
    "p = Path.cwd()\n",
    "while p != p.parent and not (p / \"pyproject.toml\").exists():\n",
    "    p = p.parent\n",
    "\n",
    "sys.path.insert(0, str(p))\n",
    "print(\"Project root:\", p)\n",
    "\n",
    "\n",
    "\n",
    "import pandas as pd\n",
    "from notebooks.consts import *\n",
    "from tauso.file_utils import read_human_genome_fasta_dict\n",
    "from tauso.consts import *\n",
    "import numpy as np\n",
    "from tauso.util import get_antisense\n",
    "import pickle\n",
    "from tauso.genome.read_human_genome import get_locus_to_data_dict\n",
    "from tauso.file_utils import read_human_genome_fasta_dict\n",
    "import RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3be683b69317538c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-15T14:26:37.676466Z",
     "start_time": "2025-11-15T14:26:37.476902Z"
    }
   },
   "outputs": [],
   "source": [
    "csv_path = NOTEBOOK_PATH / 'data' / 'data_asoptimizer_updated.csv'\n",
    "all_data = pd.read_csv(str(csv_path), low_memory=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cbe8cbdc28d5da7",
   "metadata": {},
   "source": [
    "Preprocessing ASO Experimental Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ad6b4a37d01b3e6e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-15T14:26:37.789855Z",
     "start_time": "2025-11-15T14:26:37.781017Z"
    }
   },
   "outputs": [],
   "source": [
    "from notebooks.notebook_utils import log_correction\n",
    "\n",
    "# Remove rows with missing values in the INHIBITION column\n",
    "all_data_no_nan = all_data.dropna(subset=[INHIBITION]).copy()\n",
    "# Create a new column with transformed inhibition values on a negative log scale\n",
    "log_correction(all_data_no_nan) # to avoid log 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1deba986c6a7976",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-15T14:26:37.832325Z",
     "start_time": "2025-11-15T14:26:37.824740Z"
    }
   },
   "outputs": [],
   "source": [
    "# Filter the data to include only rows where the cell line organism is human\n",
    "all_data_no_nan_human = all_data_no_nan[all_data_no_nan[CELL_LINE_ORGANISM] == 'human']\n",
    "genes = all_data_no_nan[CANONICAL_GENE].copy()\n",
    "genes_u = list(set(genes))\n",
    "# Remove non-human or negative controls from the gene list\n",
    "genes_u.remove('HBV')\n",
    "genes_u.remove('negative_control')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c4bdbcb438b1e142",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-15T14:27:13.908860Z",
     "start_time": "2025-11-15T14:26:38.718690Z"
    }
   },
   "outputs": [],
   "source": [
    "from notebooks.notebook_utils import read_cached_gene_to_data\n",
    "\n",
    "gene_to_data = read_cached_gene_to_data(genes_u)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8e2ce9132f263209",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-15T14:36:55.861784Z",
     "start_time": "2025-11-15T14:36:48.693989Z"
    }
   },
   "outputs": [],
   "source": [
    "# Filter data to keep only rows with valid gene information\n",
    "all_data_human_gene = all_data_no_nan_human[all_data_no_nan_human[CANONICAL_GENE].isin(genes_u)].copy()\n",
    "\n",
    "# Define names for new columns\n",
    "SENSE_SEQUENCE = 'sense_sequence'\n",
    "PRE_MRNA_SEQUENCE = 'pre_mrna_sequence'\n",
    "SENSE_START = 'sense_start'\n",
    "SENSE_LENGTH = 'sense_length'\n",
    "\n",
    "# Initialize new columns\n",
    "all_data_human_gene[SENSE_SEQUENCE] = \"\"\n",
    "all_data_human_gene[PRE_MRNA_SEQUENCE] = \"\"\n",
    "all_data_human_gene[SENSE_START] = np.zeros_like(all_data_human_gene[CANONICAL_GENE], dtype=int)\n",
    "all_data_human_gene[SENSE_LENGTH] = np.zeros_like(all_data_human_gene[CANONICAL_GENE], dtype=int)\n",
    "\n",
    "# Iterate over each row and compute the antisense complement and the gene's pre-mRNA\n",
    "for index, row in all_data_human_gene.iterrows():\n",
    "    gene_name = row[CANONICAL_GENE]\n",
    "\n",
    "    if gene_name not in gene_to_data:\n",
    "        continue  # Skip genes not found in genome annotation\n",
    "\n",
    "    locus_info = gene_to_data[gene_name]\n",
    "    pre_mrna = locus_info.full_mrna\n",
    "    antisense = row[SEQUENCE]\n",
    "    sense = get_antisense(antisense)\n",
    "    idx = pre_mrna.find(sense)\n",
    "\n",
    "    # Store computed sequences in new columns\n",
    "    all_data_human_gene.loc[index, SENSE_START] = idx\n",
    "    all_data_human_gene.loc[index, SENSE_LENGTH] = len(antisense)\n",
    "    all_data_human_gene.at[index, SENSE_SEQUENCE] = sense\n",
    "    all_data_human_gene.at[index, PRE_MRNA_SEQUENCE] = pre_mrna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "aeca7f9430f0d773",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-15T14:37:37.432010Z",
     "start_time": "2025-11-15T14:37:37.428357Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['index', 'ISIS', 'Target_gene', 'Cell_line', 'Density(cells/well)',\n",
       "       'Transfection', 'ASO_volume(nM)', 'Treatment_Period(hours)',\n",
       "       'Primer_probe_set', 'Sequence', 'Modification', 'Location',\n",
       "       'Chemical_Pattern', 'Linkage', 'Linkage_Location', 'Smiles',\n",
       "       'Inhibition(%)', 'seq_length', 'Canonical Gene Name',\n",
       "       'Cell line organism', 'Transcript', 'Location_in_sequence',\n",
       "       'Location_div_by_length', 'true_length_of_seq', 'mod_scan',\n",
       "       'cell_line_uniform', 'log_inhibition', 'sense_sequence',\n",
       "       'pre_mrna_sequence', 'sense_start', 'sense_length'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_data_human_gene.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21f7e043",
   "metadata": {},
   "source": [
    "Generate features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c4c3613",
   "metadata": {},
   "outputs": [],
   "source": [
    "# purine content\n",
    "from src.tauso.features.seq_features import purine_content\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "all_data_human_gene['ASO_purine_content'] = all_data_human_gene[\"Sequence\"].apply(lambda x: purine_content(x))\n",
    "sequence_features = ['ASO_purine_content']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "202335ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqUtils import gc_fraction\n",
    "\n",
    "all_data_human_gene['ASO_gc_content'] = all_data_human_gene[\"Sequence\"].apply(gc_fraction)\n",
    "sequence_features.append('ASO_gc_content')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46e46dd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# count_g_runs\n",
    "from src.tauso.features.seq_features import count_g_runs\n",
    "all_data_human_gene['ASO_ggg_counts'] = all_data_human_gene[\"Sequence\"].apply(lambda x: count_g_runs(x))\n",
    "sequence_features.append(\"ASO_ggg_counts\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06d02d57",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ENC\n",
    "from src.tauso.features.seq_features import compute_ENC\n",
    "all_data_human_gene[\"ASO_ENC\"] = all_data_human_gene[\"Sequence\"].apply(lambda x: compute_ENC(x))\n",
    "sequence_features.append(\"ASO_ENC\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5e90d92",
   "metadata": {},
   "outputs": [],
   "source": [
    "# palindromic count of 4 and 6 palindroms\n",
    "from src.tauso.features.seq_features import palindromic_fraction\n",
    "all_data_human_gene['ASO_4 palindromic'] = all_data_human_gene[SEQUENCE].apply(lambda x: palindromic_fraction(x,4))\n",
    "sequence_features.append(\"ASO_4 palindromic\")\n",
    "\n",
    "all_data_human_gene['ASO_6 palindromic'] = all_data_human_gene[SEQUENCE].apply(lambda x: palindromic_fraction(x,6))\n",
    "sequence_features.append(\"ASO_6 palindromic\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a915aeef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# entropy\n",
    "from src.tauso.features.seq_features import seq_entropy\n",
    "all_data_human_gene[\"ASO_entropy\"] = all_data_human_gene[\"Sequence\"].apply(lambda x: seq_entropy(x))\n",
    "sequence_features.append(\"ASO_entropy\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0970803d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#hairpin_score\n",
    "from src.tauso.features.seq_features import hairpin_score\n",
    "all_data_human_gene[\"ASO_hairpin_score\"] = all_data_human_gene[\"Sequence\"].apply(lambda x : hairpin_score(x))\n",
    "sequence_features.append(\"ASO_hairpin_score\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb749a68",
   "metadata": {},
   "outputs": [],
   "source": [
    "#toxic_motif_count\n",
    "from src.tauso.features.seq_features import toxic_motif_count\n",
    "all_data_human_gene[\"ASO_toxic_motif_count\"] = all_data_human_gene[\"Sequence\"].apply(lambda x : toxic_motif_count(x))\n",
    "sequence_features.append(\"ASO_toxic_motif_count\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3bbe6ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "# hairpin_dG_energy\n",
    "from src.tauso.features.seq_features import hairpin_dG_energy\n",
    "all_data_human_gene[\"ASO_hairpin_dG_energy\"] = all_data_human_gene[\"Sequence\"].apply(lambda x : hairpin_dG_energy(x))\n",
    "sequence_features.append(\"ASO_hairpin_dG_energy\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "900387a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "#hairpin_tm\n",
    "from src.tauso.features.seq_features import hairpin_tm\n",
    "all_data_human_gene[\"ASO_hairpin_tm\"] = all_data_human_gene[\"Sequence\"].apply(lambda x : hairpin_tm(x))\n",
    "sequence_features.append(\"ASO_hairpin_tm\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18b01834",
   "metadata": {},
   "outputs": [],
   "source": [
    "#tandem_repeats_score\n",
    "from src.tauso.features.seq_features import tandem_repeats_score\n",
    "all_data_human_gene[\"ASO_tandem_repeats_score\"] = all_data_human_gene[\"Sequence\"].apply(lambda x : tandem_repeats_score(x))\n",
    "sequence_features.append(\"ASO_tandem_repeats_score\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16b43c87",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dispersed_repeats_score\n",
    "from src.tauso.features.seq_features import dispersed_repeats_score\n",
    "all_data_human_gene[\"ASO_dispersed_repeats_score\"] = all_data_human_gene[\"Sequence\"].apply(lambda x : dispersed_repeats_score(x))\n",
    "sequence_features.append(\"ASO_dispersed_repeats_score\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59847f3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#flexible_dinucleotide_fraction\n",
    "from src.tauso.features.seq_features import flexible_dinucleotide_fraction\n",
    "all_data_human_gene[\"ASO_flexible_dinucleotide_fraction\"] = all_data_human_gene[\"Sequence\"].apply(lambda x : flexible_dinucleotide_fraction(x))\n",
    "sequence_features.append(\"ASO_flexible_dinucleotide_fraction\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb7a9017",
   "metadata": {},
   "outputs": [],
   "source": [
    "#gc_skew\n",
    "from src.tauso.features.seq_features import gc_skew\n",
    "all_data_human_gene[\"ASO_gc_skew\"] = all_data_human_gene[\"Sequence\"].apply(lambda x : gc_skew(x))\n",
    "sequence_features.append(\"ASO_gc_skew\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3bba0dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#gc_skew_ends\n",
    "from src.tauso.features.seq_features import gc_skew_ends\n",
    "all_data_human_gene[\"ASO_gc_skew_ends\"] = all_data_human_gene[\"Sequence\"].apply(lambda x : gc_skew_ends(x))\n",
    "sequence_features.append(\"ASO_gc_skew_ends\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03839a74",
   "metadata": {},
   "outputs": [],
   "source": [
    "#at_skew\n",
    "from src.tauso.features.seq_features import at_skew\n",
    "all_data_human_gene[\"ASO_at_skew\"] = all_data_human_gene[\"Sequence\"].apply(lambda x : at_skew(x))\n",
    "sequence_features.append(\"ASO_at_skew\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3352d0d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#stop_codon_count\n",
    "from src.tauso.features.seq_features import stop_codon_count\n",
    "all_data_human_gene[\"ASO_stop_codon_count\"] = all_data_human_gene[\"Sequence\"].apply(lambda x : stop_codon_count(x))\n",
    "sequence_features.append(\"ASO_stop_codon_count\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a189798",
   "metadata": {},
   "outputs": [],
   "source": [
    "#nucleotide_diversity\n",
    "from src.tauso.features.seq_features import nucleotide_diversity\n",
    "all_data_human_gene[\"ASO_nucleotide_diversity\"] = all_data_human_gene[\"Sequence\"].apply(lambda x : nucleotide_diversity(x))\n",
    "sequence_features.append(\"ASO_nucleotide_diversity\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "86d522a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#cg_dinucleotide_fraction\n",
    "import importlib\n",
    "#import niv_features\n",
    "#importlib.reload(niv_features)\n",
    "from src.tauso.features.seq_features import cg_dinucleotide_fraction\n",
    "all_data_human_gene[\"ASO_cg_dinucleotide_fraction\"] = all_data_human_gene[\"Sequence\"].apply(lambda x : cg_dinucleotide_fraction(x))\n",
    "sequence_features.append(\"ASO_cg_dinucleotide_fraction\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b7d3b8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#poly_pyrimidine_stretch\n",
    "from src.tauso.features.seq_features import poly_pyrimidine_stretch\n",
    "all_data_human_gene[\"ASO_poly_pyrimidine_stretch\"] = all_data_human_gene[\"Sequence\"].apply(lambda x : poly_pyrimidine_stretch(x))\n",
    "sequence_features.append(\"ASO_poly_pyrimidine_stretch\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6281b084",
   "metadata": {},
   "outputs": [],
   "source": [
    "#dinucleotide_entropy\n",
    "from src.tauso.features.seq_features import dinucleotide_entropy\n",
    "all_data_human_gene[\"ASO_dinucleotide_entropy\"] = all_data_human_gene[\"Sequence\"].apply(lambda x : dinucleotide_entropy(x))\n",
    "sequence_features.append(\"ASO_dinucleotide_entropy\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7d583a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#gc_block_length\n",
    "from src.tauso.features.seq_features import gc_block_length\n",
    "all_data_human_gene[\"ASO_gc_block_length\"] = all_data_human_gene[\"Sequence\"].apply(lambda x : gc_block_length(x))\n",
    "sequence_features.append(\"ASO_gc_block_length\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9c5228d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#at_rich_region_score\n",
    "from src.tauso.features.seq_features import at_rich_region_score\n",
    "all_data_human_gene[\"ASO_at_rich_region_score\"] = all_data_human_gene[\"Sequence\"].apply(lambda x : at_rich_region_score(x))\n",
    "sequence_features.append(\"ASO_at_rich_region_score\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bcd7c40",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.tauso.features.tai import calc_tAI, tai_weights\n",
    "weights_tai = tai_weights(\"hm\")\n",
    "all_data_human_gene[\"ASO_tai\"] = all_data_human_gene[\"Sequence\"].apply(lambda x : calc_tAI(x, weights_tai))\n",
    "sequence_features.append(\"ASO_tai\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "992dce98",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.tauso.features.seq_features import gc_content_3prime_end\n",
    "all_data_human_gene[\"ASO_gc_content_3prime_end\"] = all_data_human_gene[\"Sequence\"].apply(lambda x : gc_content_3prime_end(x))\n",
    "sequence_features.append(\"ASO_gc_content_3prime_end\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "027d12eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "from src.tauso.features.seq_features import homooligo_count\n",
    "all_data_human_gene[\"ASO_homooligo_count\"] = all_data_human_gene[\"Sequence\"].apply(lambda x : homooligo_count(x))\n",
    "sequence_features.append(\"ASO_homooligo_count\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82216e2b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "aso_design",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
