{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "IMPORTS\n",
    "==================\n",
    "\n"
   ],
   "id": "723ac7e097c1c6f4"
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "initial_id",
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "from io import StringIO\n",
    "import os\n",
    "import pickle"
   ]
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "from src.tauso.off_target.Roni.off_target_functions import dna_to_rna_reverse_complement, normalize_chrom, get_min_max_coords, parse_risearch_output, aggregate_off_targets, parse_gtf, parse_fasta, name2accession\n",
    "\n",
    "from src.tauso.off_target.Roni.get_premRNA_sequences import get_relevant_exp_data, get_premrna_coords, add_original_sequence, final_func_premrna\n",
    "\n",
    "from src.tauso.off_target.Roni.mutate_cell_line_transcriptome import get_expression_of_cell_line, get_mutations_of_cell_line, mutate, mutation_dict, celline_list, find_shift"
   ],
   "id": "c80a24716dd4c191"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "DATA\n",
    "===="
   ],
   "id": "3f40a8909c000bea"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Paths",
   "id": "231a13ab33c86952"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "PROJECT_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), \"..\", \"..\"))\n",
    "DATA_DIR = os.path.join(PROJECT_ROOT, \"scripts\", \"data_genertion\", \"cell_line_expression\")\n",
    "\n",
    "fasta_path = os.path.join(DATA_DIR, \"Homo_sapiens.GRCh38.dna_sm.primary_assembly.fa\")\n",
    "gtf_path = os.path.join(DATA_DIR, \"gencode.v48.chr_patch_hapl_scaff.annotation.gtf\")\n",
    "exp_path = os.path.join(DATA_DIR, \"OmicsExpressionProteinCodingGenesTPMLogp1.csv\")\n",
    "mut_path = os.path.join(DATA_DIR, \"OmicsSomaticMutations.csv\")"
   ],
   "id": "4624267f107452b7"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "Select the cell-line(s) you work on",
   "id": "969befb97dbf08e2"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "cell_line_lst = [\"ACH-000681\"]\n",
    "cell_line2data = {}"
   ],
   "id": "317e663da0f280e2"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "",
   "id": "825893930c0b378c"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "I. GET SEQUENCES\n",
    "="
   ],
   "id": "5e5b8874b80e13b3"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "# Step 0 - Expression and mutation data per cell line\n",
    "for cell_line in cell_line_lst:\n",
    "    exp_data = get_expression_of_cell_line(cell_line, exp_path)\n",
    "    mut_data = get_mutations_of_cell_line(cell_line, mut_path)\n",
    "    cell_line2data[cell_line] = [exp_data, mut_data]\n",
    "\n",
    "# Step 1 - Parse FASTA\n",
    "chrom2seq = parse_fasta(\"fasta_file_path\", \"fasta_pickle_output_file_path\")\n",
    "\n",
    "# Step 2 - Parse GTF\n",
    "t_id2coords = parse_gtf(\"gtf_file_path\", \"gtf_pickle_output_file_path\")\n",
    "\n",
    "# Step 3 - ???\n",
    "# =========== GET SMALLER SEQUENCES FILES (only if not enough GB RAM)\n",
    "\n",
    "# Step 4 - Extract pre-mRNA sequences\n",
    "transcriptomes = final_func_premrna(cell_line2data, chrom2seq, t_id2coords)\n"
   ],
   "id": "751f352264347f57"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "II. MUTATE SEQUENCES\n",
    "="
   ],
   "id": "8574c64f3e4e0b11"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "\n",
    "addition_exp = '_transcriptome_premRNA.merged.csv'\n",
    "\n",
    "with open(DATA_DIR + '_gtf_annotations.pkl', 'rb') as f:\n",
    "    annotations = pickle.load(f)\n",
    "print('GTF annotations loaded.')\n",
    "\n",
    "for cell_line in celline_list:\n",
    "    print(f'processing {cell_line}')\n",
    "    # Extract necessary data\n",
    "    mut_data = pd.read_csv(DATA_DIR + cell_line + '_mutations.csv')\n",
    "    exp_data = pd.read_csv(DATA_DIR + cell_line + addition_exp)\n",
    "\n",
    "    exp_data_indexed = exp_data.set_index('Transcript_ID')\n",
    "    print('Data loaded successfully.')\n",
    "\n",
    "    for idx, row in mut_data.iterrows():\n",
    "        mut_dict = mutation_dict(row)\n",
    "\n",
    "        if mut_dict is not None:\n",
    "            tid = mut_dict['id']\n",
    "            mut_idx = mut_dict['start']\n",
    "\n",
    "\n",
    "            if tid in exp_data_indexed.index:\n",
    "                try:\n",
    "                    shift = find_shift(annotations[tid], mut_idx)\n",
    "                    print(f\"{tid} shift: {shift}\")\n",
    "                    if pd.isna(exp_data_indexed.at[tid, 'Mutated Transcript Sequence']):\n",
    "                        seq = exp_data_indexed.at[tid, 'Original Transcript Sequence']\n",
    "\n",
    "                    else:\n",
    "                        seq = exp_data_indexed.at[tid, 'Mutated Transcript Sequence']\n",
    "\n",
    "                    mutated_seq = mutate(mut_dict, shift, seq)\n",
    "                    exp_data_indexed.at[tid, 'Mutated Transcript Sequence'] = mutated_seq\n",
    "                    if seq != mutated_seq:\n",
    "                        print(f'really mutated {tid}')\n",
    "                    else:\n",
    "                        print(f'didnt really do anything... check {tid}')\n",
    "                except Exception as e:\n",
    "                    print(f\"Skipping {tid} due to error: {e}\")\n",
    "\n",
    "    exp_data_indexed.to_csv(DATA_DIR + cell_line + '.mutated' + addition_exp)\n",
    "    print(f'Saved {cell_line}')"
   ],
   "id": "a1caaeba59bdca24"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
