{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "initial_id",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-15T14:26:37.473823Z",
     "start_time": "2025-11-15T14:26:36.957896Z"
    },
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Project root: /home/nivd1/TAUSO\n"
     ]
    }
   ],
   "source": [
    "# conda activate aso_design\n",
    "import sys\n",
    "from pathlib import Path\n",
    "\n",
    "p = Path.cwd()\n",
    "while p != p.parent and not (p / \"pyproject.toml\").exists():\n",
    "    p = p.parent\n",
    "\n",
    "sys.path.insert(0, str(p))\n",
    "print(\"Project root:\", p)\n",
    "\n",
    "\n",
    "\n",
    "import pandas as pd\n",
    "from notebooks.consts import *\n",
    "from tauso.file_utils import read_human_genome_fasta_dict\n",
    "from tauso.consts import *\n",
    "import numpy as np\n",
    "from tauso.util import get_antisense\n",
    "import pickle\n",
    "from tauso.genome.read_human_genome import get_locus_to_data_dict\n",
    "from tauso.file_utils import read_human_genome_fasta_dict\n",
    "import RNA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "3be683b69317538c",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-15T14:26:37.676466Z",
     "start_time": "2025-11-15T14:26:37.476902Z"
    }
   },
   "outputs": [],
   "source": [
    "csv_path = NOTEBOOK_PATH / 'data' / 'data_asoptimizer_updated.csv'\n",
    "all_data = pd.read_csv(str(csv_path), low_memory=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4cbe8cbdc28d5da7",
   "metadata": {},
   "source": [
    "Preprocessing ASO Experimental Dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ad6b4a37d01b3e6e",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-15T14:26:37.789855Z",
     "start_time": "2025-11-15T14:26:37.781017Z"
    }
   },
   "outputs": [],
   "source": [
    "from notebooks.notebook_utils import log_correction\n",
    "\n",
    "# Remove rows with missing values in the INHIBITION column\n",
    "all_data_no_nan = all_data.dropna(subset=[INHIBITION]).copy()\n",
    "# Create a new column with transformed inhibition values on a negative log scale\n",
    "log_correction(all_data_no_nan) # to avoid log 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1deba986c6a7976",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-15T14:26:37.832325Z",
     "start_time": "2025-11-15T14:26:37.824740Z"
    }
   },
   "outputs": [],
   "source": [
    "# Filter the data to include only rows where the cell line organism is human\n",
    "all_data_no_nan_human = all_data_no_nan[all_data_no_nan[CELL_LINE_ORGANISM] == 'human']\n",
    "genes = all_data_no_nan[CANONICAL_GENE].copy()\n",
    "genes_u = list(set(genes))\n",
    "# Remove non-human or negative controls from the gene list\n",
    "genes_u.remove('HBV')\n",
    "genes_u.remove('negative_control')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c4bdbcb438b1e142",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-15T14:27:13.908860Z",
     "start_time": "2025-11-15T14:26:38.718690Z"
    }
   },
   "outputs": [],
   "source": [
    "from notebooks.notebook_utils import read_cached_gene_to_data\n",
    "\n",
    "gene_to_data = read_cached_gene_to_data(genes_u)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8e2ce9132f263209",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-15T14:36:55.861784Z",
     "start_time": "2025-11-15T14:36:48.693989Z"
    }
   },
   "outputs": [],
   "source": [
    "# Filter data to keep only rows with valid gene information\n",
    "all_data_human_gene = all_data_no_nan_human[all_data_no_nan_human[CANONICAL_GENE].isin(genes_u)].copy()\n",
    "\n",
    "# Define names for new columns\n",
    "SENSE_SEQUENCE = 'sense_sequence'\n",
    "PRE_MRNA_SEQUENCE = 'pre_mrna_sequence'\n",
    "SENSE_START = 'sense_start'\n",
    "SENSE_LENGTH = 'sense_length'\n",
    "\n",
    "# Initialize new columns\n",
    "all_data_human_gene[SENSE_SEQUENCE] = \"\"\n",
    "all_data_human_gene[PRE_MRNA_SEQUENCE] = \"\"\n",
    "all_data_human_gene[SENSE_START] = np.zeros_like(all_data_human_gene[CANONICAL_GENE], dtype=int)\n",
    "all_data_human_gene[SENSE_LENGTH] = np.zeros_like(all_data_human_gene[CANONICAL_GENE], dtype=int)\n",
    "\n",
    "# Iterate over each row and compute the antisense complement and the gene's pre-mRNA\n",
    "for index, row in all_data_human_gene.iterrows():\n",
    "    gene_name = row[CANONICAL_GENE]\n",
    "\n",
    "    if gene_name not in gene_to_data:\n",
    "        continue  # Skip genes not found in genome annotation\n",
    "\n",
    "    locus_info = gene_to_data[gene_name]\n",
    "    pre_mrna = locus_info.full_mrna\n",
    "    antisense = row[SEQUENCE]\n",
    "    sense = get_antisense(antisense)\n",
    "    idx = pre_mrna.find(sense)\n",
    "\n",
    "    # Store computed sequences in new columns\n",
    "    all_data_human_gene.loc[index, SENSE_START] = idx\n",
    "    all_data_human_gene.loc[index, SENSE_LENGTH] = len(antisense)\n",
    "    all_data_human_gene.at[index, SENSE_SEQUENCE] = sense\n",
    "    all_data_human_gene.at[index, PRE_MRNA_SEQUENCE] = pre_mrna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "aeca7f9430f0d773",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-15T14:37:37.432010Z",
     "start_time": "2025-11-15T14:37:37.428357Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['index', 'ISIS', 'Target_gene', 'Cell_line', 'Density(cells/well)',\n",
       "       'Transfection', 'ASO_volume(nM)', 'Treatment_Period(hours)',\n",
       "       'Primer_probe_set', 'Sequence', 'Modification', 'Location',\n",
       "       'Chemical_Pattern', 'Linkage', 'Linkage_Location', 'Smiles',\n",
       "       'Inhibition(%)', 'seq_length', 'Canonical Gene Name',\n",
       "       'Cell line organism', 'Transcript', 'Location_in_sequence',\n",
       "       'Location_div_by_length', 'true_length_of_seq', 'mod_scan',\n",
       "       'cell_line_uniform', 'log_inhibition', 'sense_sequence',\n",
       "       'pre_mrna_sequence', 'sense_start', 'sense_length'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_data_human_gene.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6f78b03535a94b62",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-11-15T14:37:37.629198Z",
     "start_time": "2025-11-15T14:37:37.618338Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>ISIS</th>\n",
       "      <th>Target_gene</th>\n",
       "      <th>Cell_line</th>\n",
       "      <th>Density(cells/well)</th>\n",
       "      <th>Transfection</th>\n",
       "      <th>ASO_volume(nM)</th>\n",
       "      <th>Treatment_Period(hours)</th>\n",
       "      <th>Primer_probe_set</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>...</th>\n",
       "      <th>Location_in_sequence</th>\n",
       "      <th>Location_div_by_length</th>\n",
       "      <th>true_length_of_seq</th>\n",
       "      <th>mod_scan</th>\n",
       "      <th>cell_line_uniform</th>\n",
       "      <th>log_inhibition</th>\n",
       "      <th>sense_sequence</th>\n",
       "      <th>pre_mrna_sequence</th>\n",
       "      <th>sense_start</th>\n",
       "      <th>sense_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>540733</td>\n",
       "      <td>K-RAS</td>\n",
       "      <td>A431</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>free uptake</td>\n",
       "      <td>2000.0</td>\n",
       "      <td>24</td>\n",
       "      <td>RTS3496_MGB</td>\n",
       "      <td>GCTAAAACAAATGCTA</td>\n",
       "      <td>...</td>\n",
       "      <td>41212.0</td>\n",
       "      <td>0.901972</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>A431</td>\n",
       "      <td>-4.204842</td>\n",
       "      <td>TAGCATTTGTTTTAGC</td>\n",
       "      <td>GATTTTCCTAGGCGGCGGCCGCGGCGGCGGAGGCAGCAGCGGCGGC...</td>\n",
       "      <td>41212</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>540747</td>\n",
       "      <td>K-RAS</td>\n",
       "      <td>A431</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>free uptake</td>\n",
       "      <td>2000.0</td>\n",
       "      <td>24</td>\n",
       "      <td>RTS3496_MGB</td>\n",
       "      <td>TATAATGGTGAATATC</td>\n",
       "      <td>...</td>\n",
       "      <td>23686.0</td>\n",
       "      <td>0.518395</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>A431</td>\n",
       "      <td>-4.532707</td>\n",
       "      <td>GATATTCACCATTATA</td>\n",
       "      <td>GATTTTCCTAGGCGGCGGCCGCGGCGGCGGAGGCAGCAGCGGCGGC...</td>\n",
       "      <td>23686</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>540806</td>\n",
       "      <td>K-RAS</td>\n",
       "      <td>A431</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>free uptake</td>\n",
       "      <td>2000.0</td>\n",
       "      <td>24</td>\n",
       "      <td>RTS3496_MGB</td>\n",
       "      <td>GCATGAAGATTTCTGG</td>\n",
       "      <td>...</td>\n",
       "      <td>43363.0</td>\n",
       "      <td>0.949049</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>A431</td>\n",
       "      <td>-3.637849</td>\n",
       "      <td>CCAGAAATCTTCATGC</td>\n",
       "      <td>GATTTTCCTAGGCGGCGGCCGCGGCGGCGGAGGCAGCAGCGGCGGC...</td>\n",
       "      <td>43363</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>651479</td>\n",
       "      <td>K-RAS</td>\n",
       "      <td>A431</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>free uptake</td>\n",
       "      <td>2000.0</td>\n",
       "      <td>24</td>\n",
       "      <td>RTS3496_MGB</td>\n",
       "      <td>GGTGAATATCTTCAAA</td>\n",
       "      <td>...</td>\n",
       "      <td>23680.0</td>\n",
       "      <td>0.518264</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>A431</td>\n",
       "      <td>-4.276805</td>\n",
       "      <td>TTTGAAGATATTCACC</td>\n",
       "      <td>GATTTTCCTAGGCGGCGGCCGCGGCGGCGGAGGCAGCAGCGGCGGC...</td>\n",
       "      <td>23680</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>651490</td>\n",
       "      <td>K-RAS</td>\n",
       "      <td>A431</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>free uptake</td>\n",
       "      <td>2000.0</td>\n",
       "      <td>24</td>\n",
       "      <td>RTS3496_MGB</td>\n",
       "      <td>CACTTGTACTAGTATG</td>\n",
       "      <td>...</td>\n",
       "      <td>41168.0</td>\n",
       "      <td>0.901009</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>A431</td>\n",
       "      <td>-4.159039</td>\n",
       "      <td>CATACTAGTACAAGTG</td>\n",
       "      <td>GATTTTCCTAGGCGGCGGCCGCGGCGGCGGAGGCAGCAGCGGCGGC...</td>\n",
       "      <td>41168</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   index    ISIS Target_gene Cell_line  Density(cells/well) Transfection  \\\n",
       "0      0  540733       K-RAS      A431               5000.0  free uptake   \n",
       "1      1  540747       K-RAS      A431               5000.0  free uptake   \n",
       "2      2  540806       K-RAS      A431               5000.0  free uptake   \n",
       "3      3  651479       K-RAS      A431               5000.0  free uptake   \n",
       "4      4  651490       K-RAS      A431               5000.0  free uptake   \n",
       "\n",
       "   ASO_volume(nM)  Treatment_Period(hours) Primer_probe_set          Sequence  \\\n",
       "0          2000.0                       24      RTS3496_MGB  GCTAAAACAAATGCTA   \n",
       "1          2000.0                       24      RTS3496_MGB  TATAATGGTGAATATC   \n",
       "2          2000.0                       24      RTS3496_MGB  GCATGAAGATTTCTGG   \n",
       "3          2000.0                       24      RTS3496_MGB  GGTGAATATCTTCAAA   \n",
       "4          2000.0                       24      RTS3496_MGB  CACTTGTACTAGTATG   \n",
       "\n",
       "   ... Location_in_sequence Location_div_by_length true_length_of_seq  \\\n",
       "0  ...              41212.0               0.901972                 16   \n",
       "1  ...              23686.0               0.518395                 16   \n",
       "2  ...              43363.0               0.949049                 16   \n",
       "3  ...              23680.0               0.518264                 16   \n",
       "4  ...              41168.0               0.901009                 16   \n",
       "\n",
       "  mod_scan cell_line_uniform log_inhibition    sense_sequence  \\\n",
       "0        0              A431      -4.204842  TAGCATTTGTTTTAGC   \n",
       "1        0              A431      -4.532707  GATATTCACCATTATA   \n",
       "2        1              A431      -3.637849  CCAGAAATCTTCATGC   \n",
       "3        0              A431      -4.276805  TTTGAAGATATTCACC   \n",
       "4        0              A431      -4.159039  CATACTAGTACAAGTG   \n",
       "\n",
       "                                   pre_mrna_sequence sense_start sense_length  \n",
       "0  GATTTTCCTAGGCGGCGGCCGCGGCGGCGGAGGCAGCAGCGGCGGC...       41212           16  \n",
       "1  GATTTTCCTAGGCGGCGGCCGCGGCGGCGGAGGCAGCAGCGGCGGC...       23686           16  \n",
       "2  GATTTTCCTAGGCGGCGGCCGCGGCGGCGGAGGCAGCAGCGGCGGC...       43363           16  \n",
       "3  GATTTTCCTAGGCGGCGGCCGCGGCGGCGGAGGCAGCAGCGGCGGC...       23680           16  \n",
       "4  GATTTTCCTAGGCGGCGGCCGCGGCGGCGGAGGCAGCAGCGGCGGC...       41168           16  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_data_human_gene.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bf5bfd23",
   "metadata": {},
   "source": [
    "generate modification features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67cbbd13",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute_mod_fraction\n",
    "from src.tauso.features.mod_features import compute_mod_fraction\n",
    "\n",
    "all_data_human_gene['Modification_fraction'] = all_data_human_gene[\"Chemical_Pattern\"].apply(lambda x: compute_mod_fraction(x))\n",
    "modification_features = ['Modification_fraction']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc147bf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute_mod_type_count\n",
    "from src.tauso.features.mod_features import compute_mod_type_count\n",
    "\n",
    "all_data_human_gene['Modification_type_count'] = all_data_human_gene[\"Chemical_Pattern\"].apply(lambda x: compute_mod_type_count(x))\n",
    "modification_features.append('Modification_type_count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a23713d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute_mod_5prime_run\n",
    "from src.tauso.features.mod_features import compute_mod_5prime_run\n",
    "\n",
    "all_data_human_gene['Modification_5prime_run'] = all_data_human_gene[\"Chemical_Pattern\"].apply(lambda x: compute_mod_5prime_run(x))\n",
    "modification_features.append('Modification_5prime_run')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "791fae3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute_mod_3prime_run\n",
    "from src.tauso.features.mod_features import compute_mod_3prime_run\n",
    "\n",
    "all_data_human_gene['Modification_3prime_run'] = all_data_human_gene[\"Chemical_Pattern\"].apply(lambda x: compute_mod_3prime_run(x))\n",
    "modification_features.append('Modification_3prime_run')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "defda600",
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute_mod_min_distance_to_5prime\n",
    "from src.tauso.features.mod_features import compute_mod_min_distance_to_5prime\n",
    "\n",
    "all_data_human_gene['Modification_min_distance_to_5prime'] = all_data_human_gene[\"Chemical_Pattern\"].apply(lambda x: compute_mod_min_distance_to_5prime(x))\n",
    "modification_features.append('Modification_min_distance_to_5prime')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a42233f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute_mod_min_distance_to_3prime\n",
    "from src.tauso.features.mod_features import compute_mod_min_distance_to_3prime\n",
    "\n",
    "all_data_human_gene['Modification_min_distance_to_3prime'] = all_data_human_gene[\"Chemical_Pattern\"].apply(lambda x: compute_mod_min_distance_to_3prime(x))\n",
    "modification_features.append('Modification_min_distance_to_3prime')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9c53302",
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute_mod_pos_std\n",
    "from src.tauso.features.mod_features import compute_mod_pos_std\n",
    "\n",
    "all_data_human_gene['Modification_pos_std'] = all_data_human_gene[\"Chemical_Pattern\"].apply(lambda x: compute_mod_pos_std(x))\n",
    "modification_features.append('Modification_pos_std')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fc793cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute_mod_block_count\n",
    "from src.tauso.features.mod_features import compute_mod_block_count\n",
    "\n",
    "all_data_human_gene['Modification_block_count'] = all_data_human_gene[\"Chemical_Pattern\"].apply(lambda x: compute_mod_block_count(x))\n",
    "modification_features.append('Modification_block_count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ac1603d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute_mod_max_block_length\n",
    "from src.tauso.features.mod_features import compute_mod_max_block_length\n",
    "\n",
    "all_data_human_gene['Modification_max_block_length'] = all_data_human_gene[\"Chemical_Pattern\"].apply(lambda x: compute_mod_max_block_length(x))\n",
    "modification_features.append('Modification_max_block_length')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b0d6ed0",
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute_mod_char_entropy\n",
    "from src.tauso.features.mod_features import compute_mod_char_entropy\n",
    "\n",
    "all_data_human_gene['Modification_char_entropy'] = all_data_human_gene[\"Chemical_Pattern\"].apply(lambda x: compute_mod_char_entropy(x))\n",
    "modification_features.append('Modification_char_entropy')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c7361bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute_dominant_mod_fraction\n",
    "from src.tauso.features.mod_features import compute_dominant_mod_fraction\n",
    "\n",
    "all_data_human_gene['Modification_dominant_mod_fraction'] = all_data_human_gene[\"Chemical_Pattern\"].apply(lambda x: compute_dominant_mod_fraction(x))\n",
    "modification_features.append('Modification_dominant_mod_fraction')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "daf60ea9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute_mod_evenness\n",
    "from src.tauso.features.mod_features import compute_mod_evenness\n",
    "\n",
    "all_data_human_gene['Modification_evenness'] = all_data_human_gene[\"Chemical_Pattern\"].apply(lambda x: compute_mod_evenness(x))\n",
    "modification_features.append('Modification_evenness')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "254f5746",
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute_mod_symmetry_score\n",
    "from src.tauso.features.mod_features import compute_mod_symmetry_score\n",
    "\n",
    "all_data_human_gene['Modification_symmetry_score'] = all_data_human_gene[\"Chemical_Pattern\"].apply(lambda x: compute_mod_symmetry_score(x))\n",
    "modification_features.append('Modification_symmetry_score')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6df93fd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute_mod_skew_index\n",
    "from src.tauso.features.mod_features import compute_mod_skew_index\n",
    "\n",
    "all_data_human_gene['Modification_skew_index'] = all_data_human_gene[\"Chemical_Pattern\"].apply(lambda x: compute_mod_skew_index(x))\n",
    "modification_features.append('Modification_skew_index')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5549dc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute_mod_mean_gap\n",
    "from src.tauso.features.mod_features import compute_mod_mean_gap\n",
    "\n",
    "all_data_human_gene['Modification_mean_gap'] = all_data_human_gene[\"Chemical_Pattern\"].apply(lambda x: compute_mod_mean_gap(x))\n",
    "modification_features.append('Modification_mean_gap')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e0da5ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute_mod_local_density_max\n",
    "from src.tauso.features.mod_features import compute_mod_local_density_max\n",
    "\n",
    "all_data_human_gene['Modification_local_density_max'] = all_data_human_gene[\"Chemical_Pattern\"].apply(lambda x: compute_mod_local_density_max(x))\n",
    "modification_features.append('Modification_local_density_max')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a117e58",
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute_mod_in_core\n",
    "from src.tauso.features.mod_features import compute_mod_in_core\n",
    "\n",
    "all_data_human_gene['Modification_in_core'] = all_data_human_gene[\"Chemical_Pattern\"].apply(lambda x: compute_mod_in_core(x))\n",
    "modification_features.append('Modification_in_core')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4001db62",
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute_mod_longest_repeat_run\n",
    "from src.tauso.features.mod_features import compute_mod_longest_repeat_run\n",
    "\n",
    "all_data_human_gene['Modification_longest_repeat_run'] = all_data_human_gene[\"Chemical_Pattern\"].apply(lambda x: compute_mod_longest_repeat_run(x))\n",
    "modification_features.append('Modification_longest_repeat_run')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "517fd797",
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute_mod_adjacent_pair_count\n",
    "from src.tauso.features.mod_features import compute_mod_adjacent_pair_count\n",
    "\n",
    "all_data_human_gene['Modification_adjacent_pair_count'] = all_data_human_gene[\"Chemical_Pattern\"].apply(lambda x: compute_mod_adjacent_pair_count(x))\n",
    "modification_features.append('Modification_adjacent_pair_count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "614ea9eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#compute_mod_strong_repeat_group_count\n",
    "from src.tauso.features.mod_features import compute_mod_strong_repeat_group_count\n",
    "\n",
    "all_data_human_gene['Modification_strong_repeat_group_count'] = all_data_human_gene[\"Chemical_Pattern\"].apply(lambda x: compute_mod_strong_repeat_group_count(x))\n",
    "modification_features.append('Modification_strong_repeat_group_count')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "aso_design",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
